syntax = "proto3";

enum OrderSideProto {
  ORDER_SIDE_UNSPECIFIED = 0;
  ORDER_SIDE_BUY = 1;
  ORDER_SIDE_SELL = 2;
}
enum OrderTypeProto {
  ORDER_TYPE_UNSPECIFIED = 0;
  ORDER_TYPE_LIMIT = 1;
  ORDER_TYPE_MARKET = 2;
}

message OrderProto {
  uint64 id = 1;
  OrderSideProto side = 2;
  OrderTypeProto type = 3;
  string price = 4;
  string quantity = 5;
  uint64 timestamp = 6;
}

message TradeProto {
  uint64 buy_order_id = 1;
  uint64 sell_order_id = 2;
  string price = 3;
  string quantity = 4;
  uint64 timestamp = 5;
}

message SubmitOrderRequest {
  OrderProto Order = 1;
}
message SubmitOrderResponse {}

message GetTradeListRequest {}
message GetTradeListResponse {
  repeated TradeProto TradeList = 1; 
}

message CleanTradeListRequest {}
message CleanTradeListResponse {}

service EngineService {
  rpc SubmitOrder(SubmitOrderRequest) returns (SubmitOrderResponse);
  rpc GetTradeList(GetTradeListRequest) returns (GetTradeListResponse);
  rpc CleanTradeList(CleanTradeListRequest) returns (GetTradeListResponse);
}